cardshifter-server/src/test/java/com/cardshifter/server/main/ServerConnectionTest.java

```diff
@@ -252,5 +252,16 @@ public void testUsersQueryNotLoggedIn() throws IOException, InterruptedException
 		// There shouldn't be a UserStatusMessage for client2
 		assertNoMessage(client1);
 	}
+
+	@Test(timeout = 5000)
+	public void testNoOfflineMessageIfNotLoggedIn() throws IOException, InterruptedException {
+		System.out.println("testNoOfflineMessageIfNotLoggedIn");
+		TestClient client2 = createTestClient();
+		client2.disconnect();
+
+		// Avoid race condition between UserStatusMessage and ServerStatusMessage sent by assertNoMessage
+		Thread.sleep(1000);
```

Comment:
One way could be to instantiate a subclass of the `Server` that overrides the `onDisconnected` method and calls first `super.onDisconnected` and then alerts a [`CountDownLatch.countDown()`](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html) that the Test-thread waits for before calling `assertNoMessage`.

(Hey, I didn't say it was a clean way...)

As the `Server` instance is available in the test, and it is on that instance that the `onDisconnected` method is called, it should be possible to wait for that instance to have processed that before continuing.